[FLINK-9386] Embed netty router

pnowojski opened the pull request:
This replaces netty-router dependency with our own version of it, which is simplified and adds guarantees about order of matching router patterns.

This is a prerequisite for https://issues.apache.org/jira/browse/FLINK-3952 . netty-router 1.10 is incompatible with Netty 4.1, while netty-router 2.2.0 brakes a compatibility in a way that we were unable to use it. 
 
## Verifying this change

This change ports part of the netty-router code, including tests and adds a test for preserving patter matching order by the `Router`. Additionally this change is covered by various of our ITCases.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/MethodlessRouter.java:
```diff
@@ -0,0 +1,121 @@
+	private static final Logger log = LoggerFactory.getLogger(MethodlessRouter.class);
+
+	// A path pattern can only point to one target
+	private final Map<PathPattern, T> routes = new LinkedHashMap<>();
```
Please correct me if I'm wrong but I have a question regarding memory visibility: The thread that updates this map and the Netty event loop threads are different, so there might theoretically be a memory visibility issue if routes are added after the router has been passed to the `RouterHandler`. I don't think that we do this currently, but the API theoretically allows it. I'm wondering whether it makes sense to make the routes immutable, maybe something like creating the routes with a builder:
```java
Routes routes = new RoutesBuilder()
  .addGet(...)
  ...
  .build();
Router router = new Router(routes);
```
Or use a thread-safe map here that also preserves ordering (maybe wrap using `synchronizedMap`).

---
Since we currently rely on correct registration order (e.g. `/jobs/overview` before `/jobs/:id` for correct matching), the immutable approach would allow us to include a utility in `Routes` that sorts pattern as done in `RestServerEndpoint:143`:
```java
Collections.sort(handlers,RestHandlerUrlComparator.INSTANCE);
```

uce replied:
Fine with me to not invest more time into this and keep it as is 👍

uce commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/RuntimeMonitorHandler.java:
```diff
@@ -89,19 +91,20 @@ public RuntimeMonitorHandler(
-	protected void respondAsLeader(ChannelHandlerContext ctx, Routed routed, JobManagerGateway jobManagerGateway) {
+	protected void respondAsLeader(ChannelHandlerContext ctx, RoutedRequest routedRequest, JobManagerGateway jobManagerGateway) {
 		CompletableFuture<FullHttpResponse> responseFuture;
+		RouteResult result = routedRequest.getRouteResult();
```
I think if you do `RouteResult<?> result = routedRequest.getRouteResult();` you don't need the cast to `Set<String>` in lines 101 and 106.

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/PathPattern.java:
```diff
@@ -0,0 +1,179 @@
+ * <p>The pattern will be broken to tokens, example:
+ * {@code ["constant1", ":variable", "constant2", ":*"]}
+ */
+final class PathPattern {
```
Maybe add reference to https://github.com/sinetja/netty-router/blob/2.2.0/src/main/java/io/netty/handler/codec/http/router/PathPattern.java

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/RouteResult.java:
```diff
@@ -0,0 +1,136 @@
+ *
+ * <p>Result of calling {@link Router#route(HttpMethod, String)}.
+ */
+public class RouteResult<T> {
```
Maybe add reference to https://github.com/sinetja/netty-router/blob/2.2.0/src/main/java/io/netty/handler/codec/http/router/RouteResult.java

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/Router.java:
```diff
@@ -0,0 +1,398 @@
+import java.util.Set;
+
+/**
+ * This is adopted and simplified code from tv.cntt:netty-router library. Compared to original version this one
```
- Maybe add a link to https://github.com/sinetja/netty-router/blob/2.2.0/src/main/java/io/netty/handler/codec/http/router/Router.java

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/RouterHandler.java:
```diff
@@ -0,0 +1,109 @@
+			}
+		}
+
+		channelHandlerContext.fireChannelRead(new RoutedRequest(routeResult, httpRequest).retain());
```
changed into
```
RoutedRequest<?> request = new RoutedRequest<>(routeResult, httpRequest);
channelHandlerContext.fireChannelRead(request.retain());
```
I like having `.retain()` calls to be in the same line where they are actually designed for, since it's kind of self documenting which line is the reason behind retaining.

uce replied:
:+1:

pnowojski commented:
Thanks for the review and manual checks! I either addressed your comments in fixup commit and left comment responses otherwise.

> Do we file a follow up ticket to remove netty-router from our shaded Netty? Optionally, we might consider adding an import suppression for org.apache.flink.shaded.netty4.io.netty.handler.codec.http.router.* to guard against any accidental usage until we remove the dependency.

I planned to upgrade netty and drop netty-router in one step of upgrading `flink-shaded-netty`. Do you think it should be split somehow?

> I don't know what your opinion is on cleaning up the copied classes. I'm OK with keeping everything as is, but I get a bunch of minor/trivial warnings about the code (like code visibility, unchecked calls, unused variables, etc.). I understand if we want to stay close to the copied classes.

I would be open to clean it up as we think it's best. I do not see a point in maintaining compatibility with the original sources. If you have other such comments, please feel free to write them.

pnowojski commented:
Thanks :) I have squashed commits and rebased the PR on latest master.
